core/src/main/java/io/grpc/Metadata.java

```diff
@@ -235,7 +305,25 @@ public boolean containsKey(Key<?> key) {
   public <T> void put(Key<T> key, T value) {
     Preconditions.checkNotNull(key, "key");
     Preconditions.checkNotNull(value, "value");
-    storeAdd(key.name, new MetadataEntry(key, value));
+    if (len() == 0 || len() == cap()) {
+      int desiredCap = len() + (len() >>> 1) + 4;
+      expand(desiredCap & ~1); // floor to a multiple of 2
+    }
+    name(size, key.asciiName());
+    value(size, key.toBytes(value));
+    typedValues[size] = value;
+    size++;
+  }
+
+  private void expand(int newCapacity) {
+    byte[][] newNamesAndValues = new byte[newCapacity][];
+    Object[] newTypedValues = new Object[newCapacity];
+    if (size != 0) {
```

Comment:
size = 0 is a cheat.  I use size = 0 to cover the case when the array is empty and when the array is null.  I don't think arraycopy works when the input array is null, even if the size is 0, though I may be wrong.